File: backend/app/services/text_chunking_service.py
from typing import List, Dict, Any


class TextChunkingService:
    """
    Service responsible for dividing input text into manageable chunks
    for localized analysis, applying overlapping boundaries.
    """

    def chunk_text(self, text: str, chunk_size: int, overlap: int) -> List[Dict[str, Any]]:
        """
        Splits text into chunks of `chunk_size` characters with `overlap` character overlap.
        """
        if not text:
            return []

        # Validate parameters and fall back to sensible defaults
        if chunk_size <= 0:
            chunk_size = 800
        if overlap < 0 or overlap >= chunk_size:
            overlap = 100

        chunks = []
        text_len = len(text)
        start = 0
        chunk_idx = 1

        while start < text_len:
            # Determine end position
            end = min(start + chunk_size, text_len)
            chunk_content = text[start:end]

            chunks.append({
                "chunk_id": f"chunk_{chunk_idx:03d}",
                "text": chunk_content
            })

            chunk_idx += 1

            # Break if we reached the end of the text
            if end >= text_len:
                break

            # Slide window forward by (size - overlap)
            start += (chunk_size - overlap)

        return chunks


# Export a singleton instance
chunking_service = TextChunkingService()

File: backend-node/src/services/textChunkingService.ts
export interface TextChunk {
  chunk_id: string
  text: string
}

class TextChunkingService {
  chunkText(text: string, chunkSize: number, overlap: number): TextChunk[] {
    if (!text) return []

    let effectiveChunkSize = chunkSize
    let effectiveOverlap = overlap
    if (effectiveChunkSize <= 0) effectiveChunkSize = 800
    if (effectiveOverlap < 0 || effectiveOverlap >= effectiveChunkSize) effectiveOverlap = 100

    const chunks: TextChunk[] = []
    const textLen = text.length
    let start = 0
    let chunkIdx = 1

    while (start < textLen) {
      const end = Math.min(start + effectiveChunkSize, textLen)
      const chunkContent = text.slice(start, end)

      chunks.push({
        chunk_id: `chunk_${String(chunkIdx).padStart(3, '0')}`,
        text: chunkContent,
      })

      chunkIdx += 1

      if (end >= textLen) break

      start += effectiveChunkSize - effectiveOverlap
    }

    return chunks
  }
}

export const chunkingService = new TextChunkingService()
